### PR TITLE
Fire discard cards event when discarding as a cost

### LINKED ISSUE #7368: Rielle, the Everwise doesn't trigger of cycling
I cast Rielle, cycle a card but Rielle doesn't trigger

```
11:49 PM: Turn 53 copperfield (17 - 11 - 7 - 0)
11:49 PM, T53.UP: copperfield - Ability triggers: Ephara, God of the Polis [bcb] - At the beginning of each upkeep, if you had another creature enter the battlefield under your control last turn, draw a card.
11:49 PM, T53.UP: copperfield draws a card
11:49 PM, T53.UP: copperfield - Ability triggers: Ominous Seas [373] - Whenever you draw a card, put a foreshadow counter on Ominous Seas [373].
11:49 PM, T53.DR: copperfield draws a card
11:49 PM, T53.DR: copperfield - Ability triggers: Ominous Seas [373] - Whenever you draw a card, put a foreshadow counter on Ominous Seas [373].
11:49 PM, T53.DR: copperfield - Ability triggers: Gavi, Nest Warden [9e8] - Whenever you draw your second card each turn, create a 2/2 red and white Dinosaur Cat creature token.
11:49 PM, T53.DR: copperfield creates a Dinosaur Cat [fa6] token
11:50 PM, T53.M1: copperfield casts Rielle, the Everwise [c3b]
11:50 PM, T53.M1: copperfield puts Rielle, the Everwise [c3b] from stack onto the Battlefield
11:50 PM, T53.M1: copperfield puts Izzet Boilerworks [425] from hand onto the Battlefield
11:50 PM, T53.M1: copperfield plays Izzet Boilerworks [425]
11:50 PM, T53.M1: copperfield - Ability triggers: Izzet Boilerworks [425] - When Izzet Boilerworks [425] enters the battlefield, return a land you control to its owner's hand.
11:50 PM, T53.M1: copperfield puts Island [c45] from battlefield into their hand
11:50 PM, T53.M1: copperfield discards Ash Barrens [ff8] (source: Ash Barrens [ff8])
11:50 PM, T53.M1: copperfield activates: search your library for a basic land card, reveal it, and put that card into your hand. Then shuffle your library. from Ash Barrens [ff8]
11:50 PM, T53.M1: copperfield - Ability triggers: Astral Slide [bd9] - Whenever a player cycles a card, you may exile target creature. If you do, return that card to the battlefield under its owner's control at the beginning of the next end step. - targeting Oreskos Explorer [b4e]
11:50 PM, T53.M1: copperfield - Ability triggers: Crystalline Resonance [6f3] - Whenever you cycle a card, you may have have Crystalline Resonance [6f3] become a copy of another target permanent until your next turn, except it has this ability. - targeting Kalemne, Disciple of Iroas [8b7]
11:51 PM, T53.M1: copperfield moves Oreskos Explorer [b4e] from battlefield to the exile zone (source: Astral Slide [bd9])
11:51 PM, T53.M1: copperfield searches their library (source: Ash Barrens [ff8])
11:51 PM, T53.M1: copperfield puts a card from library into their hand
11:51 PM, T53.M1: copperfield reveals Plains
11:51 PM, T53.M1: copperfield's library is shuffled (source: Ash Barrens [ff8])

```

version 1.4.48V0-beta4 (build: 2020-12-28 21:03)

### PR DESCRIPTION
**Pull request by smartinsempere**
This is my first pull request to the project and I think this should word but it affects not only issue fixed. Maybe more general purpose tests could be implemented.

fixes #7368

### CODE REVIEW

**Comment by weirddan455** [comment]
You should configure git to use your email address for commits so that GitHub will give you credit if this gets merged (either your real email you signed up for GitHub with or your noreply.github.com address if you prefer to keep your real email private.)

https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address

If done correctly, you'll see your GitHub username as the author of your commit.  You'll have to then resend this commit if you want it to apply to this PR.  The following will essentially overwrite this commit after you've set your email (usually you don't want to force push but in this case I think it's the easiest):

```
git reset --soft HEAD~1
git commit
git push --force
```

**Comment by smartinsempere** [comment]
I have configured email address and my username is now the author of my commit. Thanks @weirddan455 !

File: Mage/src/main/java/mage/players/PlayerImpl.java [hunk 1]
```diff
@@ -821,7 +821,7 @@ the moment before the cost was paid (see rule 717, "Handling Illegal Actions").
         // So discard is also successful if card is moved to another zone by replacement effect!
         game.fireEvent(GameEvent.getEvent(GameEvent.EventType.DISCARDED_CARD, card.getId(), source, playerId));
 
-        if (fireFinalEvent) {
+        if (payForCost) {
             game.fireEvent(GameEvent.getEvent(GameEvent.EventType.DISCARDED_CARDS, null, source, playerId, 1));
```
Comment (JayDi85) [inline Mage/src/main/java/mage/players/PlayerImpl.java:825]: That's wrong. If you want to check event for discard cost or discard effect then use event's flag in the bugged ability. Example:
shot_210124_203358

Even more a `fireFinalEvent` was used for that special case:
shot_210124_203131

I'm found that it was changed few commits ago by error. It must be used here as `true`:
shot_210124_204653

So what you need:
* revert changes and return fire flag;
* fix default `discard(` to use true flag;
* find all `DISCARDED_CARDS` event usage and setup the event flag same way as normal event do (see screenshot above about cost/effect);
* change Rielle code to check that event's flag;
* add additional commands/checks in all of your tests:
shot_210124_204142

**Review by JayDi85 — CHANGES_REQUESTED** [review]

File: Mage/src/main/java/mage/players/PlayerImpl.java (hunk 1 above)
Comment (smartinsempere) [reply Mage/src/main/java/mage/players/PlayerImpl.java]: Seems that fixing default discard flag has solved the problem with Rielle. I've changed tests to include strict choose mode. Thanks for your help.

**Review by JayDi85 — APPROVED** [review]

**Pull request merged**